File: LeetCode/Medium/Task Scheduler/Go/taskScheduler.go
package main

import (
	"container/heap"
	"fmt"
)

type M []int

func (m M) Len() int              { return len(m) }
func (m M) Less(i, j int) bool    { return m[i] > m[j] }
func (m M) Swap(i, j int)         { m[i], m[j] = m[j], m[i] }
func (m *M) Push(val interface{}) { *m = append(*m, val.(int)) }
func (m *M) Pop() interface{} {
	val := (*m)[len(*m)-1]
	*m = (*m)[:len(*m)-1]
	return val
}

type Pair struct {
	taskCnt        int
	coolDownPeriod int
}

func leastInterval(tasks []byte, n int) int {
	size := len(tasks)
	tasksCount := make(map[byte]int)
	for i := 0; i < size; i++ {
		tasksCount[tasks[i]]++
	}

	maxHeap := &M{}
	for _, v := range tasksCount {
		heap.Push(maxHeap, v)
	}

	time := 0
	q := []Pair{}
	for len(q) > 0 || maxHeap.Len() > 0 {
		time += 1

		if maxHeap.Len() > 0 {
			top := heap.Pop(maxHeap).(int) - 1
			if top != 0 { //still the task has other parts
				q = append(q, Pair{top, time + n})
			}
		}

		if len(q) > 0 {
			front := q[0]
			if time == front.coolDownPeriod {
				q = q[1:]
				heap.Push(maxHeap, front.taskCnt)
			}
		}
	}

	return time
}

func main() {
	fmt.Println(leastInterval([]byte{'A', 'A', 'A', 'B', 'B', 'B'}, 2))
}

File: LeetCode/neetcode 150/priority queue/task_scheduler.cc
class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        int freq[26] = {0};
        priority_queue<int> pq;
        for (char t : tasks) freq[t - 'A']++;
        for (int i = 0; i < 26; i++) {
            if (freq[i] > 0) pq.push(freq[i]);
        }

        int time = 0;
        while (!pq.empty()) {
            int cycle = n + 1;
            vector<int> store;
            int task_cnt = 0;
            while (cycle-- && !pq.empty()) {
                if (pq.top() > 1) store.push_back(pq.top() - 1);
                pq.pop();
                task_cnt++;
            }

            for (int& x : store) pq.push(x);
            time += (pq.empty() ? task_cnt : n + 1);
        }
        return time;
    }
};
